### PR TITLE
:sparkles: binding controller code improvements

### LINKED ISSUE #1716: feature: consistent ordering in destination and workload resolutions
### Feature Description

In a `PlacementDecision` (`Binding`), it will be easier for users if the workload and destination lists are kept in some specified order, so that when comparing successive versions it is easier to find differences.

### Proposed Solution

Specify that they are kept in the natural order for each data structure: lexicographic for structs, golang order for strings.

### Want to contribute?

- [ ] I would like to work on this issue.

### Additional Context

_No response_

### LINKED ISSUE #1790: bug: controller::getObjectFromKey has a nil,nil return path
During a code review, it was found that this choice is confusing.

"Right I missed that. Do we want to keep getObjectFromKey that way? for now I added an `|| obj == nil` to the check."

_Originally posted by @vMaroon in https://github.com/kubestellar/kubestellar/pull/1765#discussion_r1493760593_

### LINKED ISSUE #1845: handle open comments in merged PR #1831
### Feature Description

@MikeSpreitzer left none-blocking comments to be handled in future PRs.

### Proposed Solution

Implement comments in #1831.

### Want to contribute?

- [ ] I would like to work on this issue.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by vMaroon**
## Summary
This PR:
- Implements left-over comments from #1831's review:
  - This involves documentation refactoring and general code clarity improvements
  - The GVKToGVRMapper is completely removed
    - Controller::handleObject receives object resource name alongside the object
    - This required changing informer maps (and the rest of the family) to be keyed by GVR and not GVK
- Updates BindingPolicyResolver::GenerateBinding's implementation to generate bindings with sorted workload objects and destination arrays
- Fixes Controller::getObjectFromKey to not have an unexpected return path

## Related issue(s)

Fixes #1790
Fixes #1716
Fixes #1845

### CODE REVIEW

File: pkg/binding/selectors.go [hunk 1]
```diff
@@ -93,8 +96,8 @@ func (c *Controller) updateResolutions(ctx context.Context, objIdentifier util.O
 				// this case can occur if a bindingpolicy resolution was deleted AFTER
 				// starting this iteration and BEFORE getting to the NoteObject function,
 				// which occurs if a bindingpolicy was deleted in this time-window.
-				utilruntime.HandleError(fmt.Errorf("failed to note object (identifier: %v) - %w",
-					objIdentifier, err))
+				logger.V(4).Info("skipped EnsureObjectIdentifierWithVersion for object-identifier (%v) and "+
+					"bindingpolicy %s because the latter was deleted", objIdentifier, bindingPolicy.GetName())
```
Comment (MikeSpreitzer) [inline pkg/binding/selectors.go:100]: This is a call on `Info`, so you need logging keys as well as values here.

File: pkg/binding/selectors.go [hunk 2]
```diff
@@ -93,8 +96,8 @@ func (c *Controller) updateResolutions(ctx context.Context, objIdentifier util.O
 				// this case can occur if a bindingpolicy resolution was deleted AFTER
 				// starting this iteration and BEFORE getting to the NoteObject function,
 				// which occurs if a bindingpolicy was deleted in this time-window.
```
Comment (MikeSpreitzer) [inline pkg/binding/selectors.go:98]: FYI: Now the window is even smaller, it opens after the new check above for `c.bindingPolicyResolver.ResolutionExists`.

**Comment by vMaroon** [comment]
@MikeSpreitzer addressed all comments, and added the above commit to stop repeatedly starting informers on API resources. 
The latter was observed during a test, where every update to a binding-policy would start informers on Binding and BindingPolicy APIs.

File: pkg/binding/bindingpolicy-resolution.go [hunk 3]
```diff
@@ -215,14 +215,44 @@ func bindingObjectRefSetFromBindingWorkload(bindingSpecWorkload *v1alpha1.Downsy
...
-func destinationsStringSetToDestinations(destinations sets.Set[string]) []v1alpha1.Destination {
-	dests := make([]v1alpha1.Destination, 0, len(destinations))
-	for d := range destinations {
-		dests = append(dests, v1alpha1.Destination{ClusterId: d})
+func destinationsStringSetToSortedDestinations(destinationsStringSet sets.Set[string]) []v1alpha1.Destination {
+	sortedDestiantions := make([]v1alpha1.Destination, 0, len(destinationsStringSet))
+
+	// sort destinations
+	sortedDestinationStrings := destinationsStringSet.UnsortedList()
+	slices.Sort(sortedDestinationStrings)
```
Comment (MikeSpreitzer) [inline pkg/binding/bindingpolicy-resolution.go:226]: FYI:

```suggestion
	sortedDestinationStrings := destinationsStringSet.List()
```

Comment (MikeSpreitzer) [reply pkg/binding/bindingpolicy-resolution.go]: Please let me mark my own conversations as "resolved".

Comment (vMaroon) [reply pkg/binding/bindingpolicy-resolution.go]: @MikeSpreitzer apologies, it was marked resolved automatically when I committed your suggestion through the UI here.

Comment (MikeSpreitzer) [reply pkg/binding/bindingpolicy-resolution.go]: Oops, sorry, my suggestion was buggy. `List` is a func in the `sets` package, not a method of `sets.Set`.

Comment (vMaroon) [reply pkg/binding/bindingpolicy-resolution.go]: Was wondering how I missed that method. So reverting is sufficient?

**Comment by vMaroon** [comment]
@MikeSpreitzer I've reverted the sorting change

File: pkg/binding/bindingpolicy-resolution.go (hunk 3 above)
Comment (MikeSpreitzer) [reply pkg/binding/bindingpolicy-resolution.go:226]: Reverting would work. So would changing the code to the following.

```go
	sortedDestinationStrings := sets.List(destinationsStringSet)
```

Comment (vMaroon) [reply pkg/binding/bindingpolicy-resolution.go]: Done.

File: pkg/binding/bindingpolicy-resolution.go [hunk 4]
```diff
@@ -215,14 +215,43 @@ func bindingObjectRefSetFromBindingWorkload(bindingSpecWorkload *v1alpha1.Downsy
...
 }
 
-func destinationsStringSetToDestinations(destinations sets.Set[string]) []v1alpha1.Destination {
-	dests := make([]v1alpha1.Destination, 0, len(destinations))
-	for d := range destinations {
-		dests = append(dests, v1alpha1.Destination{ClusterId: d})
+func destinationsStringSetToSortedDestinations(destinationsStringSet sets.Set[string]) []v1alpha1.Destination {
+	sortedDestiantions := make([]v1alpha1.Destination, 0, len(destinationsStringSet))
+
+	// sort destinations
```
[the remark this answers is not in the document]
Comment (vMaroon) [reply pkg/binding/bindingpolicy-resolution.go]: Done.

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm
/approve
Thanks!

**Pull request merged**